### ISSUE TITLE
Retry if no prices available when checked 

### ISSUE THREAD

**Issue by pufferfish-tech** [pufferfish-tech/octopus-agile-pi-prices#7]
@Chiny91 - I've [hacked the script](https://github.com/plasma-tech/octopus-agile-pi-prices/blob/master/store_prices.py) to incorporate a simple retry method as I found the update was failing regularly. 

I'm sure this could be made a lot better, but I'm a relative github/python noob so just learning as I go!

@pufferfish-tech - many thanks for sharing this great project!

_Originally posted by @plasma-tech in https://github.com/pufferfish-tech/octopus-agile-pi-prices/issues/4#issuecomment-628829824_

**Comment by sidewinder42** [comment]
My fix for this issue was to just a run the  update script every hour from 4pm to 10pm! It works a treat!

**Comment by Chiny91** [comment]
The problem, probably not a big problem, is that that multiple updating of the price database, makes the database multiple times bigger, and currently it is never pruned.

Having said that, I get the database at 5:nn pm and 7.nn pm, which doubles the size of the database but also works.

**Comment by beararmy** [comment]
If you're getting duplicates can you not do something like I've done [Here](https://github.com/beararmy/Agile-Octopus-Tools/blob/03aec2e2f64b4b63e38fca5cc45bcc79f953d32d/functions.php#L89) to check for pre-existance as part of the INSERT

Admittedly that's mariadb, but I think it's fairly generic SQL

**Comment by pufferfish-tech** [comment]
Yeah it's what I intend to do, I'm just totally unfamiliar with sqlite (I'm also a mysql/mariadb person). 
If anyone knows the syntax for SQLite I'll be very appreciative ;)

**Comment by pufferfish-tech** [comment]
Also you can delete the SQLite db at any time btw, just rerunning create_price_db to initialise it as per the readme. So, it growing too big isn't a massive deal - for now.

**Comment by joe-thomas2** [comment]
Could someone help me update my cron so it works as discussed above? Below is what I have, but it doesn’t seem to run every 30mins....
```

# Edit this file to introduce tasks to be run by cron.
# 
# Each task to run has to be defined through a single line
# indicating with different fields when the task will be run
# and what command to run for the task
# 
# To define the time you can provide concrete values for
# minute (m), hour (h), day of month (dom), month (mon),
# and day of week (dow) or use '*' in these fields (for 'any').
# 
# Notice that tasks will be started based on the cron's system
# daemon's notion of time and timezones.
# 
# Output of the crontab jobs (including errors) is sent through
# email to the user the crontab file belongs to (unless redirected).
# 
# For example, you can run a backup of all your user accounts
# at 5 a.m every week with:
# 0 5 * * 1 tar -zcf /var/backups/home.tgz /home/
# 
# For more information see the manual pages of crontab(5) and cron(8)
# 
# m h  dom mon dow   command
@reboot sleep 10; cd /home/$USER/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/$USER/cron.log
*/30 * * * * sleep 20; cd /home/$USER/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/$USER/cron.log
05 16 * * * cd /home/$USER/octopus-agile-pi-prices; /usr/bin/python3 store_prices.py -r H -t 24-10-01 /home/$USER/cron.log
```

**Comment by jerbzz** [comment]
Where are you setting $USER?

**Comment by joe-thomas2** [comment]
> Where are you setting $USER?

Not sure what you mean to be honest! Do I need to define that somewhere?

**Comment by jerbzz** [comment]
I wouldn't do it that way at all - I'd specify the full path without any variables - but if you do want to do it that way then there's this StackOverflow answer that's useful: https://stackoverflow.com/questions/2229825/where-can-i-set-environment-variables-that-crontab-will-use

**Comment by joe-thomas2** [comment]
Ok thanks - where it refers to $USER I've replaced that with my username.

**Comment by joe-thomas2** [comment]
Hey, so I don't think this is working, even with direct paths. I’ve created this using `crontab -e` 

Where it says "user" I've replaced this with my username after the home folder: 
```
@reboot sleep 10; cd /home/<user>/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/<user>/cron.log
*/30 * * * * sleep 20; cd /home/<user>/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/<user>/cron.log
05 16 * * * cd /home/<user>/octopus-agile-pi-prices; /usr/bin/python3 store_prices.py -r H -t 24-10-01 > /home/<user>/cron.log
```

**Comment by jerbzz** [comment]
What's in the log file, if anything?

**Comment by joe-thomas2** [comment]
cron.log is blank

**Comment by jerbzz** [comment]
What's the output of `journalctl -u cron.service`

**Comment by pufferfish-tech** [comment]
Oh joy, what have they broken in the python libraries now 🙄

Give me a shout if you need anything changing, It's been a while for both of us though.

**Comment by joe-thomas2** [comment]
> What's the output of `journalctl -u cron.service`

Dec 28 21:17:26 octopi systemd[1]: Started cron.service - Regular background program processing daemon.
Dec 28 21:17:26 octopi cron[587]: (CRON) INFO (pidfile fd = 3)
Dec 28 21:17:26 octopi cron[587]: (CRON) INFO (Running @reboot jobs)
Dec 28 21:17:27 octopi CRON[594]: pam_unix(cron:session): session opened for user joe_thomas(uid=1000) by joe_thomas(uid=0)
Dec 28 21:17:27 octopi CRON[655]: (joe_thomas) CMD (sleep 10; cd /home/joe_thomas/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/joe_thomas/cron.log)
Dec 28 21:17:37 octopi CRON[594]: (CRON) info (No MTA installed, discarding output)
Dec 28 21:17:37 octopi CRON[594]: pam_unix(cron:session): session closed for user joe_thomas
Dec 28 21:30:01 octopi CRON[1817]: pam_unix(cron:session): session opened for user joe_thomas(uid=1000) by joe_thomas(uid=0)
Dec 28 21:30:02 octopi CRON[1819]: (joe_thomas) CMD (sleep 20; cd /home/joe_thomas/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/joe_thomas/cron.log)
Dec 28 21:30:22 octopi CRON[1817]: (CRON) info (No MTA installed, discarding output)
Dec 28 21:30:22 octopi CRON[1817]: pam_unix(cron:session): session closed for user joe_thomas
Dec 28 21:33:01 octopi cron[587]: (joe_thomas) RELOAD (crontabs/joe_thomas)

**Comment by joe-thomas2** [comment]
> Oh joy, what have they broken in the python libraries now 🙄
> 
> Give me a shout if you need anything changing, It's been a while for both of us though.

Thanks! I've got it mostly working, but can't get the graph to show up (see here: https://github.com/pufferfish-tech/octopus-agile-pi-prices/issues/33#issuecomment-3694993717 ) and as above the cron job doesn't appear to be working

**Comment by joe-thomas2** [comment]
Or maybe I need up update something in `store_prices.py` ?

```
python3 store_prices.py
usage: store_prices.py [-h] --region H --tariff-type TARIFF
store_prices.py: error: the following arguments are required: --region/-r, --tariff-type/-t
```

**Comment by jerbzz** [comment]
It's telling you what's wrong - you aren't providing the correct argument to the script; you have to tell it what region and tariff you need. 

That's a totally different issue to both of your other questions today.

**Comment by joe-thomas2** [comment]
Yes I know but I'm not sure where to update that info, apologies but as mentioned I'm just starting out so any help would be welcome :) 

I assume I need to entire my region and tariff somewhere in here?

```
# let's get the region from the command line and make sure it's allowed!
parser = argparse.ArgumentParser(description=('Retrieve Octopus Agile prices'
                                              'and store in a SQLite database'),
                                 formatter_class=RawTextHelpFormatter)
parser.add_argument('--region', '-r', nargs=1, type=str, metavar='H', action='store', required=True,
                    help= """
https://en.wikipedia.org/wiki/Distribution_network_operator
A = East England
B = East Midlands
C = London
D = North Wales, Merseyside and Cheshire
E = West Midlands
F = North East England
G = North West England
P = North Scotland
N = South and Central Scotland
J = South East England
H = Southern England
K = South Wales
L = South West England
M = Yorkshire""",
                    choices = ['A','B','C','D','E','F','G','P','N','J','H','K','L','M'])
parser.add_argument('--tariff-type', '-t', nargs=1, type=str, dest='tariff', action='store', required=True, help="""
Find it by logging in at https://octopus.energy/dashboard/new/accounts/personal-details/api-access and scrolling to "Unit rates"
18-02-21 â€ The original version capped at 35p per unit
22-07-22 â€ The cap rose to 55p
22-08-31 â€ The cap was increased to 78p
VAR-22-10-19 â€ This version raised the cap to Â£1 per unit and also introduced a new formula.
FLEX-22-11-25 â€ Cap stays at Â£1 per unit but new formula only deducts 17.9p from higher unit prices.
    """)
args = parser.parse_args()
print('Selected region ' + args.region[0])
print('Selected tariff ' + args.tariff[0])
```

**Comment by jerbzz** [comment]
You don't need to touch the code. When you run the script you need to provide command line arguments with the region and tariff type. 

The key is in this line of the message you got 

`usage: store_prices.py [-h] --region H --tariff-type TARIFF`

So it would be something like

`store_prices.py --region B --tariff-type FLEX-22-11-25`

It's all about reading what it's telling you 👌🏻

**Comment by joe-thomas2** [comment]
Thank you - I ran the following and it was successful. I wonder if now this has been run, the cron will be successful??

I ask because at the top of octoprice_main_inky.py it states the following (hence why I was trying to run the store prices manually: 

`# This script *won't work* unless you have run (python3 store_prices.py) at least once in the last 'n' hours (n is variable, it updates 4pm every day)`

```
python3 store_prices.py --region H --tariff-type 24-10-01
Selected region H
Selected tariff 24-10-01
Requesting Agile prices from Octopus API...
Connected to database...
Ignoring 100 duplicate prices...
No prices were inserted - maybe we have them already or octopus are late with their update.
24 unneeded prices from the past were deleted.
```

**Comment by jerbzz** [comment]
You have run it manually so well done. I'm pretty sure your cron job is running based on the log output you shared so see how you go. 

About five years ago the maintainer merged a big PR from me which totally reworked all of the data storage code and it's very possible we didn't fully sort the comments or documentation 😅

**Comment by joe-thomas2** [comment]
So it looks like the cronjob has been running, however the inkphat isn't updating 🤔 

```
Dec 30 08:30:01 octopi CRON[4378]: pam_unix(cron:session): session opened for user joe_thomas(uid=1000) by joe_thomas(uid=0)
Dec 30 08:30:01 octopi CRON[4380]: (joe_thomas) CMD (sleep 20; cd /home/joe_thomas/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/joe_thomas/cron.log)
Dec 30 08:30:22 octopi CRON[4378]: (CRON) info (No MTA installed, discarding output)
Dec 30 08:30:22 octopi CRON[4378]: pam_unix(cron:session): session closed for user joe_thomas
Dec 30 09:00:01 octopi CRON[4469]: pam_unix(cron:session): session opened for user joe_thomas(uid=1000) by joe_thomas(uid=0)
Dec 30 09:00:01 octopi CRON[4472]: (joe_thomas) CMD (sleep 20; cd /home/joe_thomas/octopus-agile-pi-prices; /usr/bin/python3 octoprice_main_inky.py > /home/joe_thomas/cron.log)
Dec 30 09:00:21 octopi CRON[4469]: (CRON) info (No MTA installed, discarding output)
Dec 30 09:00:21 octopi CRON[4469]: pam_unix(cron:session): session closed for user joe_thomas
Dec 30 09:17:01 octopi CRON[4490]: pam_unix(cron:session): session opened for user root(uid=0) by root(uid=0)
Dec 30 09:17:01 octopi CRON[4492]: (root) CMD (cd / && run-parts --report /etc/cron.hourly)
Dec 30 09:17:01 octopi CRON[4490]: pam_unix(cron:session): session closed for user root
```

If I run ’python3 octoprice_main_inky.py’ manually it updates of course. Any ideas?

**Comment by jerbzz** [comment]
I would advise just using the full path to the script and getting rid of the `cd` command. So something like `/usr/bin/python3 /home/joe_thomas/octopus-agile-pi-prices/octoprice_main_inky.py`

Relative paths are always tricky in cron jobs

**Comment by joe-thomas2** [comment]
Hmm, still no joy, here's the response from ’journalctl -u cron.service' and nothing in the log file still: 

Dec 30 10:00:01 octopi CRON[4605]: pam_unix(cron:session): session opened for user joe_thomas(uid=1000) by joe_thomas(uid=0)
Dec 30 10:00:01 octopi CRON[4607]: (joe_thomas) CMD (sleep 20; cd /home/joe_thomas/octopus-agile-pi-prices; /usr/bin/python3 /home/joe_thomas/octopus-agile-pi-prices/octoprice_main_inky.py > /home/joe_thomas/cron.log)
Dec 30 10:17:01 octopi CRON[4623]: pam_unix(cron:session): session opened for user root(uid=0) by root(uid=0)
Dec 30 10:17:01 octopi CRON[4625]: (root) CMD (cd / && run-parts --report /etc/cron.hourly)
Dec 30 10:17:01 octopi CRON[4623]: pam_unix(cron:session): session closed for user root
Dec 30 10:30:01 octopi CRON[4636]: pam_unix(cron:session): session opened for user joe_thomas(uid=1000) by joe_thomas(uid=0)
Dec 30 10:30:01 octopi CRON[4638]: (joe_thomas) CMD (sleep 20; cd /home/joe_thomas/octopus-agile-pi-prices; /usr/bin/python3 /home/joe_thomas/octopus-agile-pi-prices/octoprice_main_inky.py > /home/joe_thomas/cron.log)

**Comment by joe-thomas2** [comment]
@pufferfish-tech I wonder if you could help a rookie with getting the cronjob to work successfully?
